### ISSUE TITLE
Use shared dictionary refreshing to prevent java.lang.OutOfMemoryError

### ISSUE THREAD

**Issue by make** [logstash-plugins/logstash-filter-translate#27]
I am using logstash to parse 22 GB of apache log divided into 8 files (apache.log, apache.log.1, ..., apache.log.7).
logstash input configuration:
```
input {
  file {
    sincedb_path => "/dev/null"
    path => "/opt/server-logs/apache.log*"
    type => "apache"
    start_position => "beginning"
    ignore_older => 8640000
  }
}
```

I am using translate filter with 150 MB dictionary file to add some details into events.
logstash filter configuration:
```
    translate {
      field => "ref_url"
      dictionary_path => "/tmp/url_details.json"
      destination => "ref_details"
      exact => true
    }
```

I have following logstash configuration
```
LS_HEAP_SIZE="2g"
```

I got following error after starting logstash:

```
{:timestamp=>"2016-04-15T11:50:11.246000+0300", :message=>"Pipeline main started"}
{:timestamp=>"2016-04-15T12:04:21.460000+0300", :message=>"Exception in pipelineworker, the pipeline stopped processing new events, please check your filter configuration and restart Logstash.", "exception"=>java.lang.OutOfMemoryError: Java heap space, "backtrace"=>["java.util.Arrays.copyOf(Arrays.java:3332)", "java.lang.AbstractStringBuilder.expandCapacity(AbstractStringBuilder.java:137)", "java.lang.AbstractStringBuilder.ensureCapacityInternal(AbstractStringBuilder.java:121)", "java.lang.AbstractStringBuilder.append(AbstractStringBuilder.java:421)", "java.lang.StringBuilder.append(StringBuilder.java:136)", "org.yaml.snakeyaml.scanner.ScannerImpl.scanFlowScalarNonSpaces(ScannerImpl.java:1867)", "org.yaml.snakeyaml.scanner.ScannerImpl.scanFlowScalar(ScannerImpl.java:1843)", "org.yaml.snakeyaml.scanner.ScannerImpl.fetchFlowScalar(ScannerImpl.java:1029)", "org.yaml.snakeyaml.scanner.ScannerImpl.fetchSingle(ScannerImpl.java:1004)", "org.yaml.snakeyaml.scanner.ScannerImpl.fetchMoreTokens(ScannerImpl.java:392)", "org.yaml.snakeyaml.scanner.ScannerImpl.checkToken(ScannerImpl.java:226)", "org.yaml.snakeyaml.parser.ParserImpl$ParseFlowMappingKey.produce(ParserImpl.java:735)", "org.yaml.snakeyaml.parser.ParserImpl.peekEvent(ParserImpl.java:158)", "org.yaml.snakeyaml.parser.ParserImpl.getEvent(ParserImpl.java:168)", "org.jruby.ext.psych.PsychParser.parse(PsychParser.java:187)", "org.jruby.ext.psych.PsychParser$INVOKER$i$parse.call(PsychParser$INVOKER$i$parse.gen)", "org.jruby.runtime.callsite.CachingCallSite.call(CachingCallSite.java:202)", "org.jruby.ast.CallTwoArgNode.interpret(CallTwoArgNode.java:59)", "org.jruby.ast.NewlineNode.interpret(NewlineNode.java:105)", "org.jruby.ast.BlockNode.interpret(BlockNode.java:71)", "org.jruby.ast.IfNode.interpret(IfNode.java:116)", "org.jruby.ast.NewlineNode.interpret(NewlineNode.java:105)", "org.jruby.evaluator.ASTInterpreter.INTERPRET_METHOD(ASTInterpreter.java:74)", "org.jruby.internal.runtime.methods.InterpretedMethod.call(InterpretedMethod.java:247)", "org.jruby.internal.runtime.methods.DefaultMethod.call(DefaultMethod.java:227)", "org.jruby.runtime.callsite.CachingCallSite.callBlock(CachingCallSite.java:211)", "org.jruby.runtime.callsite.CachingCallSite.callIter(CachingCallSite.java:222)", "org.jruby.ast.FCallTwoArgBlockNode.interpret(FCallTwoArgBlockNode.java:34)", "org.jruby.ast.NewlineNode.interpret(NewlineNode.java:105)", "org.jruby.ast.BlockNode.interpret(BlockNode.java:71)", "org.jruby.evaluator.ASTInterpreter.INTERPRET_METHOD(ASTInterpreter.java:74)", "org.jruby.internal.runtime.methods.InterpretedMethod.call(InterpretedMethod.java:225)"], :level=>:error}
{:timestamp=>"2016-04-15T12:04:21.467000+0300", :message=>"Exception in pipelineworker, the pipeline stopped processing new events, please check your filter configuration and restart Logstash.", "exception"=>java.lang.OutOfMemoryError: Java heap space, "backtrace"=>["java.lang.AbstractStringBuilder.<init>(AbstractStringBuilder.java:68)", "java.lang.StringBuilder.<init>(StringBuilder.java:89)", "org.yaml.snakeyaml.scanner.ScannerImpl.scanFlowScalar(ScannerImpl.java:1839)", "org.yaml.snakeyaml.scanner.ScannerImpl.fetchFlowScalar(ScannerImpl.java:1029)", "org.yaml.snakeyaml.scanner.ScannerImpl.fetchSingle(ScannerImpl.java:1004)", "org.yaml.snakeyaml.scanner.ScannerImpl.fetchMoreTokens(ScannerImpl.java:392)", "org.yaml.snakeyaml.scanner.ScannerImpl.checkToken(ScannerImpl.java:226)", "org.yaml.snakeyaml.parser.ParserImpl$ParseFlowMappingValue.produce(ParserImpl.java:762)", "org.yaml.snakeyaml.parser.ParserImpl.peekEvent(ParserImpl.java:158)", "org.yaml.snakeyaml.parser.ParserImpl.getEvent(ParserImpl.java:168)", "org.jruby.ext.psych.PsychParser.parse(PsychParser.java:187)", "org.jruby.ext.psych.PsychParser$INVOKER$i$parse.call(PsychParser$INVOKER$i$parse.gen)", "org.jruby.runtime.callsite.CachingCallSite.call(CachingCallSite.java:202)", "org.jruby.ast.CallTwoArgNode.interpret(CallTwoArgNode.java:59)", "org.jruby.ast.NewlineNode.interpret(NewlineNode.java:105)", "org.jruby.ast.BlockNode.interpret(BlockNode.java:71)", "org.jruby.ast.IfNode.interpret(IfNode.java:116)", "org.jruby.ast.NewlineNode.interpret(NewlineNode.java:105)", "org.jruby.evaluator.ASTInterpreter.INTERPRET_METHOD(ASTInterpreter.java:74)", "org.jruby.internal.runtime.methods.InterpretedMethod.call(InterpretedMethod.java:247)", "org.jruby.internal.runtime.methods.DefaultMethod.call(DefaultMethod.java:227)", "org.jruby.runtime.callsite.CachingCallSite.callBlock(CachingCallSite.java:211)", "org.jruby.runtime.callsite.CachingCallSite.callIter(CachingCallSite.java:222)", "org.jruby.ast.FCallTwoArgBlockNode.interpret(FCallTwoArgBlockNode.java:34)", "org.jruby.ast.NewlineNode.interpret(NewlineNode.java:105)", "org.jruby.ast.BlockNode.interpret(BlockNode.java:71)", "org.jruby.evaluator.ASTInterpreter.INTERPRET_METHOD(ASTInterpreter.java:74)", "org.jruby.internal.runtime.methods.InterpretedMethod.call(InterpretedMethod.java:225)", "org.jruby.internal.runtime.methods.DefaultMethod.call(DefaultMethod.java:219)", "org.jruby.runtime.callsite.CachingCallSite.call(CachingCallSite.java:202)", "org.jruby.ast.FCallTwoArgNode.interpret(FCallTwoArgNode.java:38)", "org.jruby.ast.LocalAsgnNode.interpret(LocalAsgnNode.java:123)"], :level=>:error}
{:timestamp=>"2016-04-15T12:04:21.475000+0300", :message=>"Exception in pipelineworker, the pipeline stopped processing new events, please check your filter configuration and restart Logstash.", "exception"=>java.lang.OutOfMemoryError: Java heap space, "backtrace"=>["java.util.Arrays.copyOf(Arrays.java:3332)", "java.lang.AbstractStringBuilder.expandCapacity(AbstractStringBuilder.java:137)", "java.lang.AbstractStringBuilder.ensureCapacityInternal(AbstractStringBuilder.java:121)", "java.lang.AbstractStringBuilder.append(AbstractStringBuilder.java:421)", "java.lang.StringBuilder.append(StringBuilder.java:136)", "org.yaml.snakeyaml.scanner.ScannerImpl.scanPlain(ScannerImpl.java:2022)", "org.yaml.snakeyaml.scanner.ScannerImpl.fetchPlain(ScannerImpl.java:1046)", "org.yaml.snakeyaml.scanner.ScannerImpl.fetchMoreTokens(ScannerImpl.java:401)", "org.yaml.snakeyaml.scanner.ScannerImpl.checkToken(ScannerImpl.java:226)", "org.yaml.snakeyaml.parser.ParserImpl$ParseFlowMappingValue.produce(ParserImpl.java:762)", "org.yaml.snakeyaml.parser.ParserImpl.peekEvent(ParserImpl.java:158)", "org.yaml.snakeyaml.parser.ParserImpl.getEvent(ParserImpl.java:168)", "org.jruby.ext.psych.PsychParser.parse(PsychParser.java:187)", "org.jruby.ext.psych.PsychParser$INVOKER$i$parse.call(PsychParser$INVOKER$i$parse.gen)", "org.jruby.runtime.callsite.CachingCallSite.call(CachingCallSite.java:202)", "org.jruby.ast.CallTwoArgNode.interpret(CallTwoArgNode.java:59)", "org.jruby.ast.NewlineNode.interpret(NewlineNode.java:105)", "org.jruby.ast.BlockNode.interpret(BlockNode.java:71)", "org.jruby.ast.IfNode.interpret(IfNode.java:116)", "org.jruby.ast.NewlineNode.interpret(NewlineNode.java:105)", "org.jruby.evaluator.ASTInterpreter.INTERPRET_METHOD(ASTInterpreter.java:74)", "org.jruby.internal.runtime.methods.InterpretedMethod.call(InterpretedMethod.java:247)", "org.jruby.internal.runtime.methods.DefaultMethod.call(DefaultMethod.java:227)", "org.jruby.runtime.callsite.CachingCallSite.callBlock(CachingCallSite.java:211)", "org.jruby.runtime.callsite.CachingCallSite.callIter(CachingCallSite.java:222)", "org.jruby.ast.FCallTwoArgBlockNode.interpret(FCallTwoArgBlockNode.java:34)", "org.jruby.ast.NewlineNode.interpret(NewlineNode.java:105)", "org.jruby.ast.BlockNode.interpret(BlockNode.java:71)", "org.jruby.evaluator.ASTInterpreter.INTERPRET_METHOD(ASTInterpreter.java:74)", "org.jruby.internal.runtime.methods.InterpretedMethod.call(InterpretedMethod.java:225)", "org.jruby.internal.runtime.methods.DefaultMethod.call(DefaultMethod.java:219)", "org.jruby.runtime.callsite.CachingCallSite.call(CachingCallSite.java:202)"], :level=>:error}
```

After some testing and monitoring with jvisualvm I found that logstash requires 8GB of memory to not to crash with OutOfMemoryError.

That seemed very unreasonable while the size of the dictionary file is only 150MB.

After taking full thread dump at the time of peak memory usage I found that there are 8 threads ([main]>worker0 - 7), all of which are refreshing the dictionary from the file concurrently.

```
"[main]>worker0" #21 daemon prio=5 os_prio=0 tid=0x00007f31c400c000 nid=0x26a2 runnable [0x00007f31d2398000]
   java.lang.Thread.State: RUNNABLE
...
        at rubyjit.LogStash::Filters::Translate$$filter_5971cf1767e1d72c473804b77d47ef4d404879201028566121.__file__(/opt/logstash/vendor/bundle/jruby/1.9/gems/logstash-filter-translate-2.1.3/lib/logstash/filters/translate.rb:141)
...
```
Memory usage when running logstash with LS_HEAP_SIZE="8g" 
![logstash_memory_usage](https://cloud.githubusercontent.com/assets/70431/14563952/08bbc0cc-032d-11e6-9f23-9eec2b127b38.png)

To improve the performance and to minimize the memory consumption the dictionary refreshing should be done by a single thread.

**Comment by pemontto** [comment]
:+1:

**Comment by andrewvc** [comment]
Thanks for reporting this. We should definitely only use one thread for the dictionary.

**Comment by andrewvc** [comment]
@make thanks for the excellent diagnostic. We've released this as a 5.0 compatible plugin. We may be creating a bridge soon to auto-backport all our plugins to 2.x. If that works out we'll do that, otherwise it may need to be manually backported.
